### ISSUE TITLE
Cannot sync videos on Xbox one, prompted to manually add server on restart

### ISSUE THREAD

**Issue by swratten** [jellyfin/jellyfin-kodi#115]
**Describe the bug**
As per #97 I'm experiencing similar behavior on Xbox One. 
Every time I restart jellyfin plugin or the Kodi app I am prompted to manually add server, I login and select server, 
I am asked to sync empty shows etc but do not see a prompt to sync all shows as I do on my other devices. 
I have gone to manage libraries and update libraries. 
I only have an All option, no TV shows or Movies option. I press okay and have no further progress from there.
I see no synced videos and am prompted to repeat the above every Kodi restart.

**System (please complete the following information):**
 - OS: Xbox One
 - Jellyfin Version: 10.3.7
 - Kodi Version: 18.4
 - Addon Version: 0.3.4

**Additional context**
I tested jellyfin server running off my MacBook last weekend on macos, installed Kodi on my Xbox one at that point with this plugin and had everything working at that point. 

Have since moved to running the jellyfin server on my Linux server on fedora 30. 
I have tested Kodi on multiple other machines with this plugin and cannot reproduce this behavior.
I'll try to get some logs off the Xbox and upload them later.

**Comment by swratten** [comment]
I've got a log here, highlights seem to be:

``` ERROR::resources\lib\database\jellyfin_db.py:161 'NoneType' object has no attribute '__getitem__' ```
seems to be the same as:
https://github.com/jellyfin/jellyfin-kodi/issues/30#issuecomment-498905860
```
ERROR::resources\lib\views.py:172 'auth.server'
                                            Traceback (most recent call last):
                                              File "resources\lib\views.py", line 169, in get_libraries
                                                libraries = self.server.jellyfin.get_media_folders()['Items']
                                              File "resources\lib\jellyfin\core\api.py", line 117, in get_media_folders
                                                return self.users("/Items")
                                              File "resources\lib\jellyfin\core\api.py", line 73, in users
                                                return self._get("Users/{UserId}%s" % handler, params)
                                              File "resources\lib\jellyfin\core\api.py", line 42, in _get
                                                return self._http("GET", handler, {'params': params})
                                              File "resources\lib\jellyfin\core\api.py", line 39, in _http
                                                return self.client.request(request)
                                              File "resources\lib\jellyfin\core\http.py", line 79, in request
                                                data = self._request(data)
                                              File "resources\lib\jellyfin\core\http.py", line 170, in _request
                                                data['url'] = "%s/emby/%s" % (self.config.data['auth.server'], data.pop('handler', ""))
                                            KeyError: 'auth.server'
2019-10-06 21:22:23.891 T:4332  NOTICE: JELLYFIN.views -> ERROR::resources\lib\views.py:699 Unable to retrieve libraries: 'auth.server'
                                            Traceback (most recent call last):
                                              File "resources\lib\views.py", line 697, in window_nodes
                                                self.media_folders = self.get_libraries()
                                              File "resources\lib\views.py", line 173, in get_libraries
                                                raise IndexError("Unable to retrieve libraries: %s" % error)
                                            IndexError: Unable to retrieve libraries: 'auth.server'
```
```
NOTICE: JELLYFIN.database -> ERROR::resources\lib\database\__init__.py:405 Failed to save credentials: [Errno 0] Error
```
```
ERROR::resources\lib\monitor.py:121 ServerId undefined.
                                            Traceback (most recent call last):
                                              File "resources\lib\monitor.py", line 105, in onNotification
                                                raise Exception("ServerId undefined.")
                                            Exception: ServerId undefined.
```
```
2019-10-06 21:25:48.937 T:2600  NOTICE: JELLYFIN.database -> DEBUG::resources\lib\database\__init__.py:182 ---<[ database: jellyfin ] 2537749291056
2019-10-06 21:25:48.945 T:2600   ERROR: Exception in thread Thread-10:
                                            Traceback (most recent call last):
                                              File "S:\Program Files\WindowsApps\XBMCFoundation.Kodi_18.3.0.70_x64__4n2hpmxwrvr6p\system\python\Lib\threading.py", line 801, in __bootstrap_inner
                                                self.run()
                                              File "U:\Users\UserMgr0\AppData\Local\Packages\XBMCFoundation.Kodi_4n2hpmxwrvr6p\LocalState\addons\plugin.video.jellyfin\resources\lib\library.py", line 98, in run
                                                if not self.startup():
                                              File "U:\Users\UserMgr0\AppData\Local\Packages\XBMCFoundation.Kodi_4n2hpmxwrvr6p\LocalState\addons\plugin.video.jellyfin\resources\lib\library.py", line 320, in startup
                                                Views().get_views()
                                              File "U:\Users\UserMgr0\AppData\Local\Packages\XBMCFoundation.Kodi_4n2hpmxwrvr6p\LocalState\addons\plugin.video.jellyfin\resources\lib\views.py", line 220, in get_views
                                                save_sync(self.sync)
                                              File "U:\Users\UserMgr0\AppData\Local\Packages\XBMCFoundation.Kodi_4n2hpmxwrvr6p\LocalState\addons\plugin.video.jellyfin\resources\lib\database\__init__.py", line 365, in save_sync
                                                outfile.write(unicode(data))
                                            IOError: [Errno 0] Error
```
[jellyfin_kodi_log.txt](https://github.com/jellyfin/jellyfin-kodi/files/3694151/jellyfin_kodi_log.txt)

**Comment by TrueTechy** [comment]
The beat way I've found to get the log off devices like the xbox is to install the kodi log up loader addon from the kodi repo under program addons then you can share the link it creates with us.

**Comment by swratten** [comment]
> The beat way I've found to get the log off devices like the xbox is to install the kodi log up loader addon from the kodi repo under program addons then you can share the link it creates with us.

Yeah thanks I found it as well thanks for the tip though :) hopefully the log will help track down whats going on

**Comment by swratten** [comment]
Note that in this session I didn't get presented with all the setup things I usually had to sync empty folders etc. But was still asked to manually add server again.

**Comment by TrueTechy** [comment]
Can you upgrade to 0.3.4, looking through your logs shows some issues that are fixed in later releases. Then do a database reset from the addon page and see if the problem persists?

**Comment by swratten** [comment]
> Can you upgrade to 0.3.4, looking through your logs shows some issues that are fixed in later releases. Then do a database reset from the addon page and see if the problem persists?

"User-Agent": "Jellyfin-Kodi/0.3.4"

**Comment by swratten** [comment]
I did a database reset earler today as well

**Comment by TrueTechy** [comment]
My apologies, I was trying to look at it on my phone which seems to have opened an old kodi log file. Looking through the log file isn't showing up anything that'd I'd expect to show the kind of problem you're having. Could you disable the Kodi debug option whilst leaving the jellyfin one on. Reset the DB again, and then setup the addon again when prompted on the restart of kodi. Then upload the log file at that point

**Comment by swratten** [comment]
> My apologies, I was trying to look at it on my phone which seems to have opened an old kodi log file. Looking through the log file isn't showing up anything that'd I'd expect to show the kind of problem you're having. Could you disable the Kodi debug option whilst leaving the jellyfin one on. Reset the DB again, and then setup the addon again when prompted on the restart of kodi. Then upload the log file at that point

All good I understand :) have followed the steps you have advised. Log seems the same to me, hopefully has something more for you https://paste.kodi.tv/amusabadov

**Comment by oddstr13** [comment]
```
2019-10-07 07:02:18.409 T:2768  NOTICE: JELLYFIN.database -> ERROR::resources\lib\database\__init__.py:405 Failed to save credentials: [Errno 0] Error
```
This looks like the relevant lone in the log - no stacktrace attached to that error unfortunately.

Probably related to the `Exception in thread Thread-10` error above.

**Comment by TrueTechy** [comment]
Looks to be a windows specific issue with switching between reading and writing modes on the same fike

**Comment by TrueTechy** [comment]
Confirmed this on an Xbox one, no library setup in addon setup flow, none shown in the "add libraries" menu. Same error: `IOError: [Errno 0] Error` showing up in logs

**Comment by TrueTechy** [comment]
So after much hunting, this looks to be a bug with the implementation of the IO module on the fall update 2017 of windows and although has now been fixed in later updates, this hasn't made it's way to the xbox version of windows it seems

**Comment by swratten** [comment]
@TrueTechy I can confirm the fixes on your Fork worked successfully :)
Thank-you all so much for your hard work and dedication! Let me know if you want me to keep the issue open or whether it is better to close. Cheers 🍻

**Comment by oddstr13** [comment]
We'll keep the issue open until fix is merged into master

**Comment by TrueTechy** [comment]
@oddstr13 yeah the branch is not ready to be merged, it partially works. Can do setup and initial sync but cannot do any library management

**Comment by swratten** [comment]
> @oddstr13 yeah the branch is not ready to be merged, it partially works. Can do setup and initial sync but cannot do any library management
> 
> Update: This is due to not being able to close the file as flushing the file to disk returns this error

I seem to still be prompted to manually add server and log back in when I restart Kodi, is that expected behavior?

**Comment by oddstr13** [comment]
No, that is not expected behavior.
Expected behavior is to stay logged in until config reset.

**Comment by swratten** [comment]
> No, that is not expected behavior.
> Expected behavior is to stay logged in until config reset.

New logfile, seems to be the same as before in terms of errors.
[jellyfin_kodi_log_08_10_19.txt](https://github.com/jellyfin/jellyfin-kodi/files/3701514/jellyfin_kodi_log_08_10_19.txt)

**Comment by swratten** [comment]
> @oddstr13 yeah the branch is not ready to be merged, it partially works. Can do setup and initial sync but cannot do any library management
> 
> Update: This is due to not being able to close the file as flushing the file to disk returns this error

@TrueTechy I've gone through my logs and noticed the same error:
```
ERROR::resources\lib\database\__init__.py:411 Failed to save credentials: [Errno 0] Error
```
So I have applied your workaround to the resolving function.
And now it seems to be throwing an error on webservice.py:
```
JELLYFIN.webservice -> ERROR::resources\lib\webservice.py:51 [Errno 10013] A
                                            Traceback (most recent call last):
                                              File "resources\lib\webservice.py", line 46, in run
                                                server = HttpServer(('127.0.0.1', PORT), requestHandler)
                                              File "S:\Program Files\WindowsApps\XBMCFoundation.Kodi_18.4.0.70_x64__4n2hpmxwrvr6p\system\python\Lib\SocketServer.py", line 417, in __init__
                                                self.server_bind()
                                              File "S:\Program Files\WindowsApps\XBMCFoundation.Kodi_18.4.0.70_x64__4n2hpmxwrvr6p\system\python\Lib\BaseHTTPServer.py", line 108, in server_bind
                                                SocketServer.TCPServer.server_bind(self)
                                              File "S:\Program Files\WindowsApps\XBMCFoundation.Kodi_18.4.0.70_x64__4n2hpmxwrvr6p\system\python\Lib\SocketServer.py", line 431, in server_bind
                                                self.socket.bind(self.server_address)
                                              File "S:\Program Files\WindowsApps\XBMCFoundation.Kodi_18.4.0.70_x64__4n2hpmxwrvr6p\system\python\Lib\socket.py", line 228, in meth
                                                return getattr(self._sock,name)(*args)
                                            error: [Errno 10013] A
```
One other difference I have noticed between my collection of logs is from that webservice function I see a set of LOG's:
```
LOG.info("--->[ webservice/%s ]", PORT)

LOG.info("---<[ webservice ]")
```
I only see the first LOG statement in my other logs, but in my latest with the exception above, I do see the second.
https://paste.kodi.tv/geguludezo

**Comment by TrueTechy** [comment]
That error from webservice is due to the xbox running kodi with restricted permission as noted in #120. This problem has stemmed from the xbox ones version of the python IO library and as such we cannot flush the files to disk which save state for things like credentials and sync state. Not sure how to proceed right now

**Comment by swratten** [comment]
> That error from webservice is due to the xbox running kodi with restricted permission as noted in #120. This problem has stemmed from the xbox ones version of the python IO library and as such we cannot flush the files to disk which save state for things like credentials and sync state. Not sure how to proceed right now

My apologies I didnt search for the error on here. Thanks for letting me know :) I've reverted to using DLNA through Kodi for now.

**Comment by cryptobank** [comment]
0.2.1 still works fine on xbox one. However I can confirm anything after that is broken.

**Comment by TrueTechy** [comment]
> 0.2.1 still works fine on xbox one. However I can confirm anything after that is broken. So what has changed since 0.2.1?

We changed to using the IO library to solve an issue where the older file handling library wasn't saving non-unicode characters properly

**Comment by cryptobank** [comment]
> > 0.2.1 still works fine on xbox one. However I can confirm anything after that is broken. So what has changed since 0.2.1?
> 
> We changed to using the IO library to solve an issue where the older file handling library wasn't saving non-unicode characters properly

What are you going to do?

**Comment by mcarlton00** [comment]
I'm in the process of reworking the handling of most of the data the addon is utilizing, but I'm starting with the two json files that are having issues being written to (`data.json` and `sync.json`).  Instead of writing these to a file, they'll be stored in the database, essentially eliminating the dependency on the bugged IO module.  I'll hopefully have the first iteration done this week, but I can't make any promises on that at this point

**Comment by o0donder0o** [comment]
#

> 0.2.1 still works fine on xbox one. However I can confirm anything after that is broken. So what has changed since 0.2.1?

I've downgraded from 3.5 to 2.1 and still getting the "select main server" pop-up when starting kodi. My progress doesn't seem be synced with this version neither. 

Did you stay on 2.1 or did you downgrade?

Using kodi 18.4 on XB0. 
Jellyfin server 10.4.1 (docker image)
Playback mode: Native

**Comment by TrueTechy** [comment]
> > 0.2.1 still works fine on xbox one. However I can confirm anything after that is broken. So what has changed since 0.2.1?
> 
> I've downgraded from 3.5 to 2.1 and still getting the "select main server" pop-up when starting kodi. My progress doesn't seem be synced with this version neither.
> 
> Did you stay on 2.1 or did you downgrade?
> 
> Using kodi 18.4 on XB0.
> Jellyfin server 10.4.1 (docker image)
> Playback mode: Native
@o0donder0o 

I've got a kodi instance running on 2.1 and it doesn't pop up the select server dialog. I've had issues with it not properly syncing or just not showing contents been watched though

**Comment by o0donder0o** [comment]
> > > 0.2.1 still works fine on xbox one. However I can confirm anything after that is broken. So what has changed since 0.2.1?
> > 
> > 
> > I've downgraded from 3.5 to 2.1 and still getting the "select main server" pop-up when starting kodi. My progress doesn't seem be synced with this version neither.
> > Did you stay on 2.1 or did you downgrade?
> > Using kodi 18.4 on XB0.
> > Jellyfin server 10.4.1 (docker image)
> > Playback mode: Native
> > @o0donder0o
> 
> I've got a kodi instance running on 2.1 and it doesn't pop up the select server dialog. I've had issues with it not properly syncing or just not showing contents been watched though

Thanks. I'll might try again, if the fixed addon is not yet released yet.

**Comment by TrueTechy** [comment]
@o0donder0o There is no fix in the latest version. Either we need to wait for microsoft to fix the xbox OS or we need to do a big rewrite to change over the using the sqlite db as storage for the data that's currently stored on files, which we're working on it just takes time of which we only have so much

**Comment by o0donder0o** [comment]
> @o0donder0o There is no fix in the latest version. Either we need to wait for microsoft to fix the xbox OS or we need to do a big rewrite to change over the using the sqlite db as storage for the data that's currently stored on files, which we're working on it just takes time of which we only have so much

Check. 

I understand Jellyfin is open-source and appreciate all the hard work put into the development within any financial support. Current issues are minor in my opion. Just switched from Plex free and I do have to say i'm very happy with the Jellyfin so far. 

One questions about 2.1 though. 
"I've had issues with it not properly syncing or just not showing contents been watched though"; does this mean you got this fixed on 2.1 or were the problems intermittent? If you got it fixed what did you do the resolve it.

**Comment by TrueTechy** [comment]
@o0donder0o the bug which causes this issue only on xbox's was introduced in 2.2 as the result of fixing another bug since we switched the module we use to write files to disk so before 2.1 works on the xbox

**Comment by o0donder0o** [comment]
@TrueTechy reinstalling Kodi with 2.1 did the trick. Progress seems to be synced at least from the server to Kodi on the Xbox. Connect pop-ups are gone as well.

Go try it more upcoming days and see if progress is synced 2-way.

Thanks for your help!

**Comment by oddstr13** [comment]
This issue should be fixed with v0.3.6, please try it out!

**Comment by cryptobank** [comment]
Confirming the fix is working on Xbox one x
